Preserve SrcSpan during compilation

stu204146-kiel opened the pull request:
### Issue

The `SrcSpan`s are preserved as much as possible for better error messages.
Closes #34.

### Description of the Change
I have changed the code of the features such that the source-span information is preserved during the features.

### Verification Process

I tried to test several programs over which I ran the compiler.

### Additional Notes
Maybe guard elimination contains some bugs.
I have added a `SrcSpan`-attribute to `GExp`.

stu204146-kiel committed:
Add easy cases of SrcSpan preservation #34

stu204146-kiel committed:
Resolve merge conflicts #34

stu204146-kiel committed:
Add more source span preservations #34

stu204146-kiel committed:
format code #34

stu204767 commented on src/lib/HST/Application.hs:
```diff
@@ -83,13 +83,14 @@ useAlgo :: (Members '[Env a, Fresh, GetOpt, Report] r, S.EqAST a)
   nExp' <- ifM (getOpt optOptimizeCase) (optimize nExp) (return nExp)
   return
-    $ S.Match S.NoSrcSpan name nVars (S.UnGuardedRhs S.NoSrcSpan nExp') Nothing
+    $ S.Match S.NoSrcSpan name nVars (S.UnGuardedRhs rhsSpan nExp') Nothing
```
I think the source span of the corresponding `FunBind` should be used here, both for the `Match` and the `UnGuardedRhs`. This `rhsSpan` seems to belong to the first match of the original function only, while we want to span all matches.

stu204767 commented on src/lib/HST/CoreAlgorithm.hs:
```diff
@@ -194,9 +194,10 @@ createAltsForMissingCons x cs er = mapM (createAltForMissingCon x er) cs
+    let p    | conEntryIsInfix conEntry = S.PInfixApp (S.getSrcSpan pat)
+               (head nvars) (conEntryName conEntry) (nvars !! 1)
+             | otherwise = S.PApp (S.getSrcSpan pat) (conEntryName conEntry)
+               nvars
```
If I backtracked the code correctly, then this `S.Pat` is (always?) generated in the `freshVarPat` function in the `HST.Effect.Fresh` module, where it gets `S.NoSrcSpan` as its source span.
While I do think that there is no perfect source span for a generated case alternative, there probably exists a better solution, possibly using the equations (`Eqs`) that occur in most other functions of this module.

stu204767 commented on src/lib/HST/CoreAlgorithm.hs:
```diff
@@ -306,30 +307,29 @@ computeAlt _ _ _ []
         listCon  = S.Special S.NoSrcSpan $ S.ConsCon S.NoSrcSpan
-    decomposeConPat (S.PInfixApp S.NoSrcSpan n listCon (S.PList S.NoSrcSpan nv))
-decomposeConPat (S.PTuple _ bxd ps)         = do
+    decomposeConPat (S.PInfixApp nspan n listCon (S.PList S.NoSrcSpan nv))
```
I think we should use the source span of the entire list and not the one for the first element here. We don't have a good source span for the remaining list, so we should use the source span of the entire list again in my opinion.
```suggestion
    decomposeConPat (S.PInfixApp s n listCon (S.PList s nv))
```

stu204767 commented on src/lib/HST/Feature/CaseCompletion.hs:
```diff
@@ -74,26 +74,27 @@ completeCase _ e@(S.Lit _ _)              = return e
                -> Bool
                -> Sem r (S.Exp a)
-completeLambda ps e insideLet = do
+completeLambda s ps e insideLet = do
```
Nice, adding an additional parameter like in this case would help removing some of the remaining `NoSrcSpan`s too.

stu204767 commented on src/lib/HST/Feature/GuardElimination.hs:
```diff
@@ -11,16 +11,19 @@ import qualified HST.Frontend.Syntax as S
 -- | A pair of patterns to match and a right-hand side to use when all
 --   patterns match.
-data GExp a = GExp { gExpPats :: [S.Pat a], gExpRhs :: S.Rhs a }
+data GExp a = GExp { gExpSrcSpan :: S.SrcSpan a
+                   , gExpPats    :: [S.Pat a]
+                   , gExpRhs     :: S.Rhs a
+                   }
```
Nice.

stu204767 commented on src/lib/HST/Feature/GuardElimination.hs:
```diff
@@ -194,14 +197,14 @@ applyGEMatches ms | any hasGuards ms = return <$> applyGE ms
   return
-    $ S.Match S.NoSrcSpan name varPats (S.UnGuardedRhs S.NoSrcSpan expr')
-    Nothing
+    $ S.Match lhsSpan name varPats (S.UnGuardedRhs S.NoSrcSpan expr') Nothing
```
Just as in the `useAlgo` function in the `Application` module, you could pass the source span of the `FunBind` to this function and use it here for the `Match` and the `UnGuardedRhs`.

stu204767 reviewed (changes requested):
There are some more `S.NoSrcSpan`s left that probably can be easily avoided.
Also, you apparently haven't yet looked at some of the other modules containing unnecessary `S.NoSrcSpan`s, especially `HST.Feature.Optimization`.

stu204146-kiel committed:
Apply review suggestions #34

stu204146-kiel committed:
Fix Haddock and Hlint suggestions #34

just95 commented on src/lib/HST/Effect/Fresh.hs:
```diff
@@ -40,18 +44,41 @@ makeSem ''Fresh
 freshName :: Member Fresh r => String -> Sem r (S.Name a)
 freshName prefix = S.Ident S.NoSrcSpan <$> freshIdent prefix
```
You could express the old functions in terms of their new `*WithSpan` variations.

```suggestion
freshName = flip freshNameWithSpan S.NoSrcSpan
```

I would also consider to call them `*WithSrcSpan` for consistency.

just95 commented on src/lib/HST/Feature/Optimization.hs:
```diff
@@ -152,12 +153,14 @@ renameAll ((from, to) : r) e = do
 --   of matched patterns for the scrutinee in the environment.
 addAndOpt :: Members '[PatternStack a, Fresh, Report] r
-          => S.QName a
+          => S.SrcSpan a
+          -> S.SrcSpan a
```
Add documentation to these parameters to make clear which source span is which.

just95 committed:
Add HLint configuration to restrict usage of `NoSrcSpan` #34

just95 committed:
Combine source spans of guarded right-hand sides #34

In GHC-lib only individual guarded right-hand sides have a
source span. The source span for `S.GuardedRhss` can be
derived by combining the individual source spans to the
smallest source span that spans all characters in the
individual source spans.

stu204146-kiel committed:
Apply review suggestions #34

just95 committed:
Remove empty identifiers from GHC-lib transformation #34

Empty identifiers were created during the GHC-lib
transformation because matches are used to represent
rules of function declarations as well as all other
occurences of pattern matching in GHC-lib. In order
to reduce code duplication, the transformation of
lambda and `case` expressions uses transformation
for function matches with empty function names.
This hack was removed without introducing code
duplication by introducing a data type that represents
a partially transformed match and contains no name
but GHC-lib's context field.

just95 committed:
Apply HLint suggestions #34

just95 committed:
Format code #34

stu204146-kiel committed:
Add comments #34

stu204767 commented on src/lib/HST/Effect/Fresh.hs:
```diff
@@ -38,19 +42,43 @@ makeSem ''Fresh
+freshName = flip freshNameWithSrcSpan S.NoSrcSpan
+
+-- | Generates a name for a fresh variable and sets the `S.SrcSpan` to the given
+--   value.
```
```suggestion
-- | Generates a name for a fresh variable and sets the 'S.SrcSpan' to the
--   given value.
```

stu204767 commented on src/lib/HST/Effect/Fresh.hs:
```diff
@@ -38,19 +42,43 @@ makeSem ''Fresh
+freshQName = flip freshQNameWithSrcSpan S.NoSrcSpan
+
+-- | Generates an unqualified name for a fresh variable and sets the `S.SrcSpan`
+--   to the given value.
```
```suggestion
-- | Generates an unqualified name for a fresh variable and sets the
--   'S.SrcSpan' to the given value.
```

stu204767 commented on src/lib/HST/Effect/Fresh.hs:
```diff
@@ -38,19 +42,43 @@ makeSem ''Fresh
+freshVar = flip freshVarWithSrcSpan S.NoSrcSpan
+
+-- | Generates a fresh variable expression and sets the `S.SrcSpan` to the given
+--   value.
```
```suggestion
-- | Generates a fresh variable expression and sets the 'S.SrcSpan' to the
--   given value.
```

stu204767 commented on src/lib/HST/Effect/Fresh.hs:
```diff
@@ -38,19 +42,43 @@ makeSem ''Fresh
-freshVarPat prefix = S.PVar S.NoSrcSpan <$> freshName prefix
+freshVarPat = flip freshVarPatWithSrcSpan S.NoSrcSpan
+
+-- | Generates a fresh variable pattern which `S.SrcSpan` is identical to the
```
```suggestion
-- | Generates a fresh variable pattern whose 'S.SrcSpan' is identical to the
```

stu204767 commented on src/lib/HST/Feature/Optimization.hs:
```diff
@@ -152,12 +153,14 @@ renameAll ((from, to) : r) e = do
 --   of matched patterns for the scrutinee in the environment.
 addAndOpt :: Members '[PatternStack a, Fresh, Report] r
-          => S.QName a
+          => S.SrcSpan a  -- ^ The source span for the whole case-expression
+          -> S.SrcSpan a  -- ^ The source span for the matched expression
```
```suggestion
          => S.SrcSpan a  -- ^ The source span for the whole case expression.
          -> S.SrcSpan a  -- ^ The source span for the matched expression.
```

stu204767 reviewed (changes requested):
There are still three suggestions left from my last review. Have you looked at them?

stu204767 committed:
Apply simple suggestions from my own code review

stu204767 committed:
Apply simple suggestions from my own code review

just95 committed:
Apply suggestions from code review #52

Co-authored-by: stu204767 <63296935+stu204767@users.noreply.github.com>

just95 committed:
Write documentation for `AnyMatch` #52

just95 committed:
Make imports explicit #52

just95 committed:
Remove `InfixMatch` constructor #52

stu204767 committed:
Fix typo in `src/lib/HST/Frontend/GHC/Util/AnyMatch.hs`

just95 committed:
Format code #52

stu204767 committed:
Merge pull request #52 from FreeProving/issue-34-tests

Add linter rules to restrict usage of `NoSrcSpan`

stu204146-kiel committed:
Update src/lib/HST/CoreAlgorithm.hs

Co-authored-by: stu204767 <63296935+stu204767@users.noreply.github.com>

stu204146-kiel committed:
Implement review suggestions

stu204767 commented on src/lib/HST/Application.hs:
```diff
@@ -84,10 +83,10 @@ useAlgo :: (Members '[Env a, Fresh, GetOpt, Report] r, S.EqAST a)
-      isInfix = all S.matchIsInfix ms
-  nVars <- replicateM arity (freshVarPat genericFreshPrefix)
+  let name     = S.matchName (head ms)
+      srcSpans = (map S.getSrcSpan) . S.matchPats . head $ ms
```
Shouldn't these parantheses be redundant?
```suggestion
      srcSpans = map S.getSrcSpan . S.matchPats . head $ ms
```

stu204767 reviewed (changes requested):
Okay, this seems to address one of the remaining suggestions I had, but what about [this one](https://github.com/FreeProving/haskell-src-transformations/pull/46#discussion_r477304108)?
Otherwise, you still have to merge the `main` branch into this one.

stu204146-kiel committed:
Apply HLint suggestions #34

stu204146-kiel committed:
Correct SrcSpan of guard eliminated functions #34

stu204146-kiel committed:
Correct SrcSpan of list patterns #34

stu204146-kiel committed:
Resolve merge conflicts #34

stu204146-kiel committed:
Format code #34

stu204146-kiel committed:
Apply Hlint suggestions #34

stu204146-kiel committed:
Format code #34

stu204146-kiel committed:
Fix non exhaustive pattern matching
